contracts/permissionless/OrderbookReserve.sol

```diff
@@ -120,10 +144,11 @@ contract OrderbookReserve is OrderIdManager, Utils2, KyberReserveInterface, Orde
 
         uint128 userRemainingSrcQty = uint128(srcQty);
         uint128 totalUserDstAmount = 0;
+        uint maxOrders = limits.maxOrdersPerTrade;
 
         for (
             (orderId, orderData.isLastOrder) = list.getFirstOrder();
-            ((userRemainingSrcQty > 0) && !orderData.isLastOrder);
+            ((userRemainingSrcQty > 0) && (!orderData.isLastOrder) && (maxOrders-- > 0));
```

Comment:
notice how the loop works.
1. we take order data inside the loop and work on the order.
2. we go to this statement. if its last order we shouldn't continue... since it is the last.